### ISSUE TITLE
Unsatisfied Link Error in Kafka with binary release 

### ISSUE THREAD

**Issue by KarstenSchnitter** [luben/zstd-jni#113]
Hi, 

I try to use zstd compression in my Kafka installation. I use an Ubunutu xenial system but I am somewhat limited in the installation of libraries. Therefore, I tried to use the binary releases from https://repo1.maven.org/maven2/com/github/luben/zstd-jni/. I replaced `kafka/lib/zstd-jni-1.4.0-1.jar`with `zstd-jni-1.4.0-1-linux_amd64.jar` and verified, that this jar is actually referenced by Kafka during start-up. When I try to produce messages using the zstd compression I get the following error:

```
$ kafka-console-producer.sh --broker-list $KAFKA_SERVERS --topic test-minimal --compression-codec zstd
>Hello World!
org.apache.kafka.common.KafkaException: java.lang.UnsatisfiedLinkError: no zstd-jni in java.library.path: [/usr/java/packages/lib, /usr/lib64, /lib64, /lib, /usr/lib]
Unsupported OS/arch, cannot find /linux/amd64/libzstd-jni.so or load zstd-jni from system libraries. Please try building from source the jar or providing libzstd-jni in your system.
        at org.apache.kafka.common.record.CompressionType$5.wrapForOutput(CompressionType.java:124)
        at org.apache.kafka.common.record.MemoryRecordsBuilder.<init>(MemoryRecordsBuilder.java:130)
        at org.apache.kafka.common.record.MemoryRecordsBuilder.<init>(MemoryRecordsBuilder.java:166)
        at org.apache.kafka.common.record.MemoryRecords.builder(MemoryRecords.java:534)
        at org.apache.kafka.common.record.MemoryRecords.builder(MemoryRecords.java:516)
        at org.apache.kafka.common.record.MemoryRecords.builder(MemoryRecords.java:464)
        at org.apache.kafka.clients.producer.internals.RecordAccumulator.recordsBuilder(RecordAccumulator.java:245)
        at org.apache.kafka.clients.producer.internals.RecordAccumulator.append(RecordAccumulator.java:222)
        at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:917)
        at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:856)
        at kafka.tools.ConsoleProducer$.send(ConsoleProducer.scala:75)
        at kafka.tools.ConsoleProducer$.main(ConsoleProducer.scala:57)
        at kafka.tools.ConsoleProducer.main(ConsoleProducer.scala)
Caused by: java.lang.UnsatisfiedLinkError: no zstd-jni in java.library.path: [/usr/java/packages/lib, /usr/lib64, /lib64, /lib, /usr/lib]
Unsupported OS/arch, cannot find /linux/amd64/libzstd-jni.so or load zstd-jni from system libraries. Please try building from source the jar or providing libzstd-jni in your system.
        at java.base/java.lang.ClassLoader.loadLibrary(Unknown Source)
        at java.base/java.lang.Runtime.loadLibrary0(Unknown Source)
        at java.base/java.lang.System.loadLibrary(Unknown Source)
        at com.github.luben.zstd.util.Native.load(Native.java:108)
        at com.github.luben.zstd.util.Native.load(Native.java:60)
        at com.github.luben.zstd.ZstdInputStream.<clinit>(ZstdInputStream.java:21)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Unknown Source)
        at org.apache.kafka.common.record.CompressionType.findConstructor(CompressionType.java:226)
        at org.apache.kafka.common.record.CompressionType.access$100(CompressionType.java:37)
        at org.apache.kafka.common.record.CompressionType$ZstdConstructors.<clinit>(CompressionType.java:218)
        at org.apache.kafka.common.record.CompressionType$5.wrapForOutput(CompressionType.java:122)
        ... 12 more
```
A similar message would be found in the Kafka broker logs, if I publish messages by an external producer capable of zstd compression. How can I get the zstd-jni setup correct to use zstd compression?

Additional information on versions:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.6 LTS
Release:        16.04
Codename:       xenial
$ ldd --version
ldd (Ubuntu GLIBC 2.23-0ubuntu11) 2.23
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
$ java -version
openjdk version "11.0.4" 2019-07-16
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.4+11)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.4+11, mixed mode)
$ kafka-configs.sh --version
2.3.0 (Commit:fc1aaa116b661c8a)
```

Best Regards

**Comment by luben** [comment]
What if you try `zstd-jni-1.4.0-1.jar` instead? Downloaded from Maven

**Comment by KarstenSchnitter** [comment]
`zstd-jni-1.4.0-1.jar` is provided by the Kafka 2.3.0 release. It produces the same exception, which was why I tried to replace it with `zstd-jni-1.4.0-1-linux_amd64.jar` in the hope to resolve the issue.

**Comment by luben** [comment]
Hi,
Can you try the attached jar - it's newer version but the only relevant change is to add more extensive logging of the failures to load the native code.

[zstd-jni-1.4.4-4-linux_amd64.jar.gz](https://github.com/luben/zstd-jni/files/4023302/zstd-jni-1.4.4-4-linux_amd64.jar.gz)

**Comment by KarstenSchnitter** [comment]
It still fails, here the output:
```
org.apache.kafka.common.KafkaException: java.lang.UnsatisfiedLinkError: /tmp/libzstd-jni7029200920829469535.so: /tmp/libzstd-jni7029200920829469535.so: failed to map segment from shared object               [3/273]
no zstd-jni in java.library.path: [/usr/java/packages/lib, /usr/lib64, /lib64, /lib, /usr/lib]                                                                                                                        
Unsupported OS/arch, cannot find /linux/amd64/libzstd-jni.so or load zstd-jni from system libraries. Please try building from source the jar or providing libzstd-jni in your system.                                 
Unsupported OS/arch, cannot find /linux/amd64/libzstd-jni.so or load zstd-jni from system libraries. Please try building from source the jar or providing libzstd-jni in your system.                                 
        at org.apache.kafka.common.record.CompressionType$5.wrapForOutput(CompressionType.java:124)
        at org.apache.kafka.common.record.MemoryRecordsBuilder.<init>(MemoryRecordsBuilder.java:130)
        at org.apache.kafka.common.record.MemoryRecordsBuilder.<init>(MemoryRecordsBuilder.java:166)
        at org.apache.kafka.common.record.MemoryRecords.builder(MemoryRecords.java:534)
        at org.apache.kafka.common.record.MemoryRecords.builder(MemoryRecords.java:516)
        at org.apache.kafka.common.record.MemoryRecords.builder(MemoryRecords.java:464)
        at org.apache.kafka.clients.producer.internals.RecordAccumulator.recordsBuilder(RecordAccumulator.java:245)
        at org.apache.kafka.clients.producer.internals.RecordAccumulator.append(RecordAccumulator.java:222)
        at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:917)
        at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:856)
        at kafka.tools.ConsoleProducer$.send(ConsoleProducer.scala:75)
        at kafka.tools.ConsoleProducer$.main(ConsoleProducer.scala:57)
        at kafka.tools.ConsoleProducer.main(ConsoleProducer.scala)
Caused by: java.lang.UnsatisfiedLinkError: /tmp/libzstd-jni7029200920829469535.so: /tmp/libzstd-jni7029200920829469535.so: failed to map segment from shared object
no zstd-jni in java.library.path: [/usr/java/packages/lib, /usr/lib64, /lib64, /lib, /usr/lib]
Unsupported OS/arch, cannot find /linux/amd64/libzstd-jni.so or load zstd-jni from system libraries. Please try building from source the jar or providing libzstd-jni in your system.
Unsupported OS/arch, cannot find /linux/amd64/libzstd-jni.so or load zstd-jni from system libraries. Please try building from source the jar or providing libzstd-jni in your system.
        at java.base/java.lang.ClassLoader.loadLibrary(Unknown Source)
        at java.base/java.lang.Runtime.loadLibrary0(Unknown Source)
        at java.base/java.lang.System.loadLibrary(Unknown Source)
        at com.github.luben.zstd.util.Native.load(Native.java:108)
        at com.github.luben.zstd.util.Native.load(Native.java:60)
        at com.github.luben.zstd.ZstdInputStream.<clinit>(ZstdInputStream.java:21)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Unknown Source)
        at org.apache.kafka.common.record.CompressionType.findConstructor(CompressionType.java:226)
        at org.apache.kafka.common.record.CompressionType.access$100(CompressionType.java:37)
        at org.apache.kafka.common.record.CompressionType$ZstdConstructors.<clinit>(CompressionType.java:218)
        at org.apache.kafka.common.record.CompressionType$5.wrapForOutput(CompressionType.java:122)
        ... 12 more
```

**Comment by luben** [comment]
May be your /tmp is mounted with noexec option. If you don't want to change it, there is Native.load(tmpdir) method you can use with some other dir that does not have the same restrictions

**Comment by KarstenSchnitter** [comment]
Hi, thanks for your help so far. Indeed `/tmp` was mounted with noexec option. I changed the configuration with `-Djava.io.tmpdir=/some/other/path`, that is not mounted with noexec. The exception in the Kafka producer goes away with that. Now I get a similar exception in the Kafka broker, that accepts the write:
```
org.apache.kafka.common.KafkaException: java.lang.ExceptionInInitializerError: Cannot unpack libzstd-jni
        at org.apache.kafka.common.record.CompressionType$5.wrapForInput(CompressionType.java:133)
        at org.apache.kafka.common.record.DefaultRecordBatch.compressedIterator(DefaultRecordBatch.java:257)
        at org.apache.kafka.common.record.DefaultRecordBatch.iterator(DefaultRecordBatch.java:324)
        at scala.collection.convert.Wrappers$JIterableWrapper.iterator(Wrappers.scala:55)
        at scala.collection.IterableLike.foreach(IterableLike.scala:74)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
        at kafka.log.LogValidator$.$anonfun$validateMessagesAndAssignOffsetsCompressed$1(LogValidator.scala:269)
        at kafka.log.LogValidator$.$anonfun$validateMessagesAndAssignOffsetsCompressed$1$adapted(LogValidator.scala:261)
        at scala.collection.Iterator.foreach(Iterator.scala:941)
        at scala.collection.Iterator.foreach$(Iterator.scala:941)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
        at scala.collection.IterableLike.foreach(IterableLike.scala:74)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
        at kafka.log.LogValidator$.validateMessagesAndAssignOffsetsCompressed(LogValidator.scala:261)
        at kafka.log.LogValidator$.validateMessagesAndAssignOffsets(LogValidator.scala:73)
        at kafka.log.Log.liftedTree1$1(Log.scala:881)
        at kafka.log.Log.$anonfun$append$2(Log.scala:868)
        at kafka.log.Log.maybeHandleIOException(Log.scala:2065)
        at kafka.log.Log.append(Log.scala:850)
        at kafka.log.Log.appendAsLeader(Log.scala:819)
        at kafka.cluster.Partition.$anonfun$appendRecordsToLeader$1(Partition.scala:772)
        at kafka.utils.CoreUtils$.inLock(CoreUtils.scala:253)
        at kafka.utils.CoreUtils$.inReadLock(CoreUtils.scala:259)
        at kafka.cluster.Partition.appendRecordsToLeader(Partition.scala:759)
        at kafka.server.ReplicaManager.$anonfun$appendToLocalLog$2(ReplicaManager.scala:763)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
        at scala.collection.mutable.HashMap.$anonfun$foreach$1(HashMap.scala:149)
        at scala.collection.mutable.HashTable.foreachEntry(HashTable.scala:237)
        at scala.collection.mutable.HashTable.foreachEntry$(HashTable.scala:230)
        at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:44)
        at scala.collection.mutable.HashMap.foreach(HashMap.scala:149)
        at scala.collection.TraversableLike.map(TraversableLike.scala:237)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:230)
        at scala.collection.AbstractTraversable.map(Traversable.scala:108)
        at kafka.server.ReplicaManager.appendToLocalLog(ReplicaManager.scala:751)
        at kafka.server.ReplicaManager.appendRecords(ReplicaManager.scala:492)
        at kafka.server.KafkaApis.handleProduceRequest(KafkaApis.scala:544)
        at kafka.server.KafkaApis.handle(KafkaApis.scala:113)
        at kafka.server.KafkaRequestHandler.run(KafkaRequestHandler.scala:69)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.ExceptionInInitializerError: Cannot unpack libzstd-jni
        at com.github.luben.zstd.util.Native.load(Native.java:122)
        at com.github.luben.zstd.util.Native.load(Native.java:60)
        at com.github.luben.zstd.ZstdInputStream.<clinit>(ZstdInputStream.java:21)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Unknown Source)
        at org.apache.kafka.common.record.CompressionType.findConstructor(CompressionType.java:226)
        at org.apache.kafka.common.record.CompressionType.access$100(CompressionType.java:37)
        at org.apache.kafka.common.record.CompressionType$ZstdConstructors.<clinit>(CompressionType.java:218)
        at org.apache.kafka.common.record.CompressionType$5.wrapForInput(CompressionType.java:131)
        ... 41 more
```
It runs with the jar provided by you and also the changed java.io.tmpdir without noexec option.

**Comment by luben** [comment]
Can you check if tmpdir is writable by the JVM process. I have to add more detailed error message but it triggers on mostly permissions and io errors

**Comment by luben** [comment]
Hi @KarstenSchnitter, here is a new JAR for testing - it should give you more detailed error that would hopefully let us diagnose the issue:
[zstd-jni-1.4.4-4.linux_amd64.jar.gz](https://github.com/luben/zstd-jni/files/4056761/zstd-jni-1.4.4-4.linux_amd64.jar.gz)

**Comment by KarstenSchnitter** [comment]
Many thanks, for your help. Your previous comment already provided the solution for me, but I could only test this today. I did not have a tmpdir, that was writable for the Kafka process, only for the producer. I adopted the configuration of my vm. Now it works with the `zstd-jni-1.4.0-1.jar` provided by Kafka. So the final outcome of this issue is:

Users of the zstd-jni should make sure, that `-Djava.io.tmpdir`points to a directory, that is writable by the java process and is mounted without the `noexec` flag.
